File: frontend/src/api/payout.ts
import api from "./client";

// Errors from /api/payout carry a stable code plus optional details.
export class PayoutApiError extends Error {
  constructor(public code: string, message: string, public details: Record<string, any> = {}) {
    super(message);
  }
}

function unwrap<T>(p: Promise<any>): Promise<T> {
  return p.then(
    (r) => r.data.data as T,
    (err) => {
      const e = err?.response?.data?.error;
      if (e) throw new PayoutApiError(e.code, e.message, e.details || {});
      throw new PayoutApiError("NETWORK", err?.message || "网络错误");
    }
  );
}

export interface Bootstrap {
  has_payer: boolean;
  payer_name: string;
  has_trade_password: boolean;
  debit_symbol: string;
  min_amount: string;
  max_amount: string;
  currency: string;
}

export interface Country {
  code: string;
  name_zh: string;
  name_en: string;
}

export interface Options {
  countries: Country[];
  genders: string[];
  doc_types: string[];
  expiry_types: string[];
  employment_status: string[];
  sources_of_income: string[];
}

export interface Recipient {
  id: number;
  name: string;
  given_name: string;
  family_name: string;
  bank_name: string;
  swift_code: string;
  account_last4: string;
  currency: string;
  bank_country: string;
  complete: boolean;
  last_error: string | null;
}

export interface PayerInput {
  given_name: string;
  family_name: string;
  name_local: string;
  date_of_birth: string;
  nationality: string;
  gender: string;
  email: string;
  calling_code: string;
  phone: string;
  doc_type: string;
  doc_number: string;
  doc_issuing_country: string;
  doc_issued_date: string;
  doc_expiry_type: string;
  doc_expiry_date: string;
  identity_file_id: string;
  identity_file_name: string;
  addr_country: string;
  addr_province: string;
  addr_city: string;
  addr_line1: string;
  addr_line2: string;
  addr_postal_code: string;
  address_proof_file_id: string;
  address_proof_file_name: string;
  employment_status: string;
  source_of_income: string;
  occupation: string;
}

export interface BankInput {
  bank_country: string;
  bank_name: string;
  swift_code: string;
  account_number: string;
  iban: string;
}

export interface RecipientInput {
  given_name: string;
  family_name: string;
  date_of_birth: string;
  nationality: string;
  email: string;
  calling_code: string;
  phone: string;
  addr_country: string;
  addr_state: string;
  addr_city: string;
  addr_line1: string;
  addr_line2: string;
  addr_postal_code: string;
  bank: BankInput;
}

export type Stage =
  | "quoting"
  | "awaiting_confirm"
  | "quote_failed"
  | "paid"
  | "processing"
  | "completed"
  | "failed"
  | "canceled"
  | "refunding"
  | "refunded"
  | "unknown";

export interface Quote {
  quote_no: string;
  debit_coin: string;
  debit_total: string;
  pay_amount: string;
  fixed_fee: string;
  transaction_fee: string;
  exchange_fee: string;
  fee_currency: string;
  destination_amount: string;
  destination_currency: string;
  valid_until: number;
  kyc_url?: string;
}

export interface QuoteState {
  status: number;
  stage: Stage;
  quote: Quote | null;
  expires_in: number;
}

export interface CreatedOrder {
  id: number;
  status: number;
  stage: Stage;
  information_to_improve: any;
}

export interface OrderSummary {
  id: number;
  amount: string;
  currency: string;
  debit_coin: string;
  debit_total: string;
  recipient_name: string;
  status: number;
  stage: Stage;
  created_at: string;
}

export interface OrderDetail {
  id: number;
  third_order_no: string;
  order_no: string | null;
  amount: string;
  currency: string;
  debit_coin: string;
  usage: string;
  note: string;
  remit_method: string;
  status: number;
  stage: Stage;
  quote: Quote | null;
  exchange_rate: string;
  recipient: Recipient;
  timeline: { status: number; stage: Stage; at: string; source: string }[];
  last_error: string | null;
  created_at: string;
  confirmed_at: string | null;
}

export const payoutApi = {
  bootstrap: () => unwrap<Bootstrap>(api.get("/payout/bootstrap")),
  options: () => unwrap<Options>(api.get("/payout/options")),
  upload: (file: File) => {
    const fd = new FormData();
    fd.append("file", file);
    return unwrap<{ file_id: string; file_name: string }>(api.post("/payout/files", fd));
  },
  createPayer: (input: PayerInput) => unwrap<{ name: string }>(api.post("/payout/payer", input)),
  recipients: () => unwrap<Recipient[]>(api.get("/payout/recipients")),
  createRecipient: (input: RecipientInput) => unwrap<Recipient>(api.post("/payout/recipients", input)),
  retryRecipient: (id: number, bank: BankInput) =>
    unwrap<Recipient>(api.post(`/payout/recipients/${id}/retry`, bank)),
  setTradePassword: (password: string) => unwrap<{}>(api.post("/payout/trade-password", { password })),
  createOrder: (input: { recipient_id: number; amount: string; usage: string; note: string }) =>
    unwrap<CreatedOrder>(api.post("/payout/orders", input)),
  orders: (beforeId?: number) =>
    unwrap<{ orders: OrderSummary[]; has_more: boolean }>(
      api.get("/payout/orders", { params: beforeId ? { before_id: beforeId } : {} })
    ),
  order: (id: string | number) => unwrap<OrderDetail>(api.get(`/payout/orders/${id}`)),
  quote: (id: string | number) => unwrap<QuoteState>(api.get(`/payout/orders/${id}/quote`)),
  confirm: (id: string | number, quote_no: string, trade_password: string) =>
    unwrap<{ status: number; stage: Stage }>(api.post(`/payout/orders/${id}/confirm`, { quote_no, trade_password })),
  cancel: (id: string | number) => unwrap<{ status: number; stage: Stage }>(api.post(`/payout/orders/${id}/cancel`)),
  requote: (id: string | number) => unwrap<CreatedOrder>(api.post(`/payout/orders/${id}/requote`)),
};

// ---- display helpers ----

export const DOC_TYPE_LABEL: Record<string, string> = {
  passport: "护照",
  national_id: "身份证",
  driving_license: "驾照",
  residence_permit: "居留许可",
};

export const GENDER_LABEL: Record<string, string> = { Male: "男", Female: "女" };

export const EMPLOYMENT_LABEL: Record<string, string> = {
  employed: "受雇",
  self_employed: "自雇",
  unemployed: "待业",
  student: "学生",
  retired: "退休",
  homemaker: "家庭主妇/主夫",
  other: "其他",
};

export const INCOME_LABEL: Record<string, string> = {
  Salary: "工资/薪金收入",
  Pension: "养老金",
  Investment: "投资收益",
  Property: "房产收入",
  FriendsAndFamily: "亲友资助",
  Benefits: "福利补贴",
};

export const USAGE_PRESETS = ["Capital transfer", "Family support", "Salary", "Tuition", "Goods purchase"];

export const STAGE_LABEL: Record<Stage, string> = {
  quoting: "报价中",
  awaiting_confirm: "待确认",
  quote_failed: "报价失败",
  paid: "支付已收到",
  processing: "银行处理中",
  completed: "转账已转出",
  failed: "汇款失败",
  canceled: "已取消",
  refunding: "退款处理中",
  refunded: "已退款",
  unknown: "未知状态",
};

// User-facing message for known error codes; falls back to the server message.
export function errorText(e: unknown): string {
  if (!(e instanceof PayoutApiError)) return (e as Error)?.message || "操作失败";
  switch (e.code) {
    case "VALIDATION":
      return `请检查「${e.details.field ?? ""}」：${e.message.split(": ").slice(1).join(": ")}`;
    case "AMOUNT_OUT_OF_RANGE":
      return `汇出金额需在 ${e.details.min} ~ ${e.details.max} 之间`;
    case "FILE_EXPIRED":
      return "上传的文件已过期，请重新上传";
    case "FILE_TOO_LARGE":
      return "文件不能超过 10MB";
    case "QUOTE_EXPIRED":
      return "报价已过期，请重新获取报价";
    case "QUOTE_CHANGED":
      return "报价已变化，请刷新后重试";
    case "TRADE_PASSWORD_INVALID":
      return `交易密码错误，还可尝试 ${e.details.remaining_attempts} 次`;
    case "TRADE_PASSWORD_LOCKED":
      return `错误次数过多，请 ${Math.ceil((e.details.retry_after ?? 0) / 60)} 分钟后再试`;
    case "PAYOUT_NOT_ENABLED":
      return "当前账户未开通汇款功能";
    case "UPSTREAM":
      return `汇款服务返回错误：${e.message}${e.details.upay_code ? `（${e.details.upay_code}）` : ""}`;
    default:
      return e.message;
  }
}

export function fmtTime(s: string | number | null | undefined): string {
  if (!s) return "";
  const d = new Date(s);
  const p = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())} ${p(d.getHours())}:${p(d.getMinutes())}:${p(d.getSeconds())}`;
}

File: backend/internal/handler/payout.go
package handler

import (
	"encoding/json"
	"errors"
	"log"
	"net/http"
	"strconv"
	"time"

	"chldemo/internal/middleware"
	"chldemo/internal/repo"
	"chldemo/internal/service"
	"chldemo/internal/upayopen"

	"github.com/labstack/echo/v4"
)

// PayoutHandler serves /api/payout/* (TECH-DESIGN-汇款功能.md §7). All routes
// sit behind JWT + the payout_enabled whitelist.
type PayoutHandler struct {
	svc  *service.PayoutService
	repo *repo.Repo
}

func NewPayoutHandler(svc *service.PayoutService, r *repo.Repo) *PayoutHandler {
	return &PayoutHandler{svc: svc, repo: r}
}

type payoutErrBody struct {
	Code    string         `json:"code"`
	Message string         `json:"message"`
	Details map[string]any `json:"details,omitempty"`
}

// payoutFail renders service errors with their status/code; anything else is 500.
func payoutFail(c echo.Context, err error) error {
	var pe *service.PayoutError
	if errors.As(err, &pe) {
		return c.JSON(pe.HTTP, map[string]any{"data": nil,
			"error": payoutErrBody{Code: pe.Code, Message: pe.Msg, Details: pe.Extra}})
	}
	if errors.Is(err, repo.ErrNotFound) {
		return fail(c, http.StatusNotFound, "NOT_FOUND", "not found")
	}
	log.Printf("payout: %s %s: %v", c.Request().Method, c.Path(), err)
	return fail(c, http.StatusInternalServerError, "INTERNAL", "internal error")
}

func pathID(c echo.Context) (int64, error) {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil || id <= 0 {
		return 0, &service.PayoutError{HTTP: http.StatusNotFound, Code: "NOT_FOUND", Msg: "not found"}
	}
	return id, nil
}

// ---- bootstrap / options ----

func (h *PayoutHandler) Bootstrap(c echo.Context) error {
	ctx, uid := c.Request().Context(), middleware.UserID(c)
	payer, err := h.repo.GetPayoutPayer(ctx, uid)
	if err != nil && !errors.Is(err, repo.ErrNotFound) {
		return payoutFail(c, err)
	}
	hasPwd, err := h.svc.HasTradePassword(ctx, uid)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, map[string]any{
		"has_payer":          payer.PayerNo != "",
		"payer_name":         payer.Name,
		"has_trade_password": hasPwd,
		"debit_symbol":       h.svc.DebitSymbol(),
		"min_amount":         service.MinPayoutAmount,
		"max_amount":         service.MaxPayoutAmount,
		"currency":           "USD",
	})
}

func (h *PayoutHandler) Options(c echo.Context) error {
	areas, err := h.svc.Countries(c.Request().Context())
	if err != nil {
		return payoutFail(c, err)
	}
	countries := make([]map[string]string, 0, len(areas))
	for _, a := range areas {
		if a.Alpha2 != "" {
			countries = append(countries, map[string]string{"code": a.Alpha2, "name_zh": a.NameZh, "name_en": a.NameEn})
		}
	}
	return ok(c, map[string]any{
		"countries":         countries,
		"genders":           service.Genders,
		"doc_types":         service.DocTypes,
		"expiry_types":      service.ExpiryTypes,
		"employment_status": service.EmploymentStatus,
		"sources_of_income": service.SourcesOfIncome,
	})
}

// ---- files ----

func (h *PayoutHandler) Upload(c echo.Context) error {
	fh, err := c.FormFile("file")
	if err != nil {
		return payoutFail(c, &service.PayoutError{HTTP: http.StatusUnprocessableEntity, Code: "VALIDATION", Msg: "file: required"})
	}
	if fh.Size > service.MaxUploadBytes {
		return payoutFail(c, &service.PayoutError{HTTP: http.StatusRequestEntityTooLarge, Code: "FILE_TOO_LARGE",
			Msg: "file must be at most 10 MB"})
	}
	f, err := fh.Open()
	if err != nil {
		return payoutFail(c, err)
	}
	defer f.Close()
	id, err := h.svc.UploadFile(c.Request().Context(), f, fh.Filename)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, map[string]any{"file_id": id, "file_name": fh.Filename})
}

// ---- payer ----

func (h *PayoutHandler) GetPayer(c echo.Context) error {
	p, err := h.repo.GetPayoutPayer(c.Request().Context(), middleware.UserID(c))
	if errors.Is(err, repo.ErrNotFound) {
		return ok(c, map[string]any{"exists": false})
	} else if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, map[string]any{"exists": true, "name": p.Name})
}

func (h *PayoutHandler) CreatePayer(c echo.Context) error {
	var in service.PayerInput
	if err := c.Bind(&in); err != nil {
		return fail(c, http.StatusBadRequest, "BAD_REQUEST", "invalid body")
	}
	p, err := h.svc.CreatePayer(c.Request().Context(), middleware.UserID(c), in)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, map[string]any{"name": p.Name})
}

// ---- recipients ----

func recipientView(p repo.PayoutRecipient) map[string]any {
	return map[string]any{
		"id": p.ID, "name": p.Name(), "given_name": p.GivenName, "family_name": p.FamilyName,
		"bank_name": p.BankName, "swift_code": p.SwiftCode, "account_last4": p.AccountLast4,
		"currency": p.Currency, "bank_country": p.BankCountry, "complete": p.Complete(),
		"last_error": p.LastError, "created_at": p.CreatedAt,
	}
}

func (h *PayoutHandler) ListRecipients(c echo.Context) error {
	list, err := h.repo.ListPayoutRecipients(c.Request().Context(), middleware.UserID(c))
	if err != nil {
		return payoutFail(c, err)
	}
	out := make([]map[string]any, 0, len(list))
	for _, p := range list {
		out = append(out, recipientView(p))
	}
	return ok(c, out)
}

func (h *PayoutHandler) CreateRecipient(c echo.Context) error {
	var in service.RecipientInput
	if err := c.Bind(&in); err != nil {
		return fail(c, http.StatusBadRequest, "BAD_REQUEST", "invalid body")
	}
	p, err := h.svc.CreateRecipient(c.Request().Context(), middleware.UserID(c), in)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, recipientView(p))
}

func (h *PayoutHandler) RetryRecipient(c echo.Context) error {
	id, err := pathID(c)
	if err != nil {
		return payoutFail(c, err)
	}
	var in service.BankInput
	if err := c.Bind(&in); err != nil {
		return fail(c, http.StatusBadRequest, "BAD_REQUEST", "invalid body")
	}
	p, err := h.svc.RetryRecipientBank(c.Request().Context(), middleware.UserID(c), id, in)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, recipientView(p))
}

// ---- trade password ----

func (h *PayoutHandler) SetTradePassword(c echo.Context) error {
	var in struct {
		Password string `json:"password"`
	}
	if err := c.Bind(&in); err != nil {
		return fail(c, http.StatusBadRequest, "BAD_REQUEST", "invalid body")
	}
	if err := h.svc.SetTradePassword(c.Request().Context(), middleware.UserID(c), in.Password); err != nil {
		return payoutFail(c, err)
	}
	return ok(c, map[string]any{})
}

// ---- orders ----

func (h *PayoutHandler) CreateOrder(c echo.Context) error {
	var in struct {
		RecipientID int64  `json:"recipient_id"`
		Amount      string `json:"amount"`
		Usage       string `json:"usage"`
		Note        string `json:"note"`
	}
	if err := c.Bind(&in); err != nil {
		return fail(c, http.StatusBadRequest, "BAD_REQUEST", "invalid body")
	}
	res, err := h.svc.CreateOrder(c.Request().Context(), middleware.UserID(c), in.RecipientID, in.Amount, in.Usage, in.Note)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, createdView(res))
}

func createdView(res service.CreatedOrder) map[string]any {
	return map[string]any{"id": res.Order.ID, "status": res.Order.Status, "stage": service.Stage(res.Order.Status),
		"information_to_improve": res.Improve}
}

func (h *PayoutHandler) Requote(c echo.Context) error {
	id, err := pathID(c)
	if err != nil {
		return payoutFail(c, err)
	}
	res, err := h.svc.Requote(c.Request().Context(), middleware.UserID(c), id)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, createdView(res))
}

func (h *PayoutHandler) Quote(c echo.Context) error {
	id, err := pathID(c)
	if err != nil {
		return payoutFail(c, err)
	}
	st, err := h.svc.Quote(c.Request().Context(), middleware.UserID(c), id)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, st)
}

func (h *PayoutHandler) Confirm(c echo.Context) error {
	id, err := pathID(c)
	if err != nil {
		return payoutFail(c, err)
	}
	var in struct {
		QuoteNo       string `json:"quote_no"`
		TradePassword string `json:"trade_password"`
	}
	if err := c.Bind(&in); err != nil {
		return fail(c, http.StatusBadRequest, "BAD_REQUEST", "invalid body")
	}
	status, err := h.svc.Confirm(c.Request().Context(), middleware.UserID(c), id, in.QuoteNo, in.TradePassword)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, map[string]any{"status": status, "stage": service.Stage(status)})
}

func (h *PayoutHandler) Cancel(c echo.Context) error {
	id, err := pathID(c)
	if err != nil {
		return payoutFail(c, err)
	}
	status, err := h.svc.Cancel(c.Request().Context(), middleware.UserID(c), id)
	if err != nil {
		return payoutFail(c, err)
	}
	return ok(c, map[string]any{"status": status, "stage": service.Stage(status)})
}

func (h *PayoutHandler) ListOrders(c echo.Context) error {
	ctx, uid := c.Request().Context(), middleware.UserID(c)
	before, _ := strconv.ParseInt(c.QueryParam("before_id"), 10, 64)
	limit, _ := strconv.Atoi(c.QueryParam("limit"))
	if limit <= 0 || limit > 50 {
		limit = 20
	}
	orders, more, err := h.repo.ListPayoutOrders(ctx, uid, before, limit)
	if err != nil {
		return payoutFail(c, err)
	}
	names := map[int64]string{}
	if recs, err := h.repo.ListPayoutRecipients(ctx, uid); err == nil {
		for _, r := range recs {
			names[r.ID] = r.Name()
		}
	}
	out := make([]map[string]any, 0, len(orders))
	for _, o := range orders {
		out = append(out, map[string]any{
			"id": o.ID, "amount": o.Amount, "currency": o.Currency, "debit_coin": o.DebitCoin,
			"debit_total": snapshot(o).DebitTotal, "recipient_name": names[o.RecipientID],
			"status": o.Status, "stage": service.Stage(o.Status), "created_at": o.CreatedAt,
		})
	}
	return ok(c, map[string]any{"orders": out, "has_more": more})
}

func snapshot(o repo.PayoutOrder) service.QuoteView {
	var q service.QuoteView
	if len(o.Quote) > 0 {
		_ = json.Unmarshal(o.Quote, &q)
	}
	return q
}

type timelineEntry struct {
	Status int       `json:"status"`
	Stage  string    `json:"stage"`
	At     time.Time `json:"at"`
	Source string    `json:"source"`
}

func (h *PayoutHandler) GetOrder(c echo.Context) error {
	id, err := pathID(c)
	if err != nil {
		return payoutFail(c, err)
	}
	ctx, uid := c.Request().Context(), middleware.UserID(c)
	o, err := h.repo.GetPayoutOrder(ctx, uid, id)
	if err != nil {
		return payoutFail(c, err)
	}
	rec, err := h.repo.GetPayoutRecipient(ctx, uid, o.RecipientID)
	if err != nil {
		return payoutFail(c, err)
	}
	events, err := h.repo.ListPayoutOrderEvents(ctx, o.ID)
	if err != nil {
		return payoutFail(c, err)
	}
	timeline := make([]timelineEntry, 0, len(events))
	for _, e := range events {
		timeline = append(timeline, timelineEntry{Status: e.Status, Stage: service.Stage(e.Status), At: e.ObservedAt, Source: e.Source})
	}
	var info struct {
		RemitInfo struct {
			ExchangeRate upayopen.Str `json:"exchangeRate"`
		} `json:"remitInfo"`
	}
	if len(o.LastInfo) > 0 {
		_ = json.Unmarshal(o.LastInfo, &info)
	}
	var quote any
	if len(o.Quote) > 0 {
		quote = snapshot(o)
	}
	return ok(c, map[string]any{
		"id": o.ID, "third_order_no": o.ThirdOrderNo, "order_no": o.OrderNo,
		"amount": o.Amount, "currency": o.Currency, "debit_coin": o.DebitCoin,
		"usage": o.Usage, "note": o.Note, "remit_method": "swift",
		"status": o.Status, "stage": service.Stage(o.Status),
		"quote": quote, "exchange_rate": info.RemitInfo.ExchangeRate.String(),
		"recipient": recipientView(rec), "timeline": timeline,
		"last_error": o.LastError, "created_at": o.CreatedAt, "confirmed_at": o.ConfirmedAt,
	})
}
